File: camera/c_plus_plus/color_tracker.cpp
#include <opencv2/opencv.hpp>
#include <opencv2/core/core.hpp>
#include <opencv2/highgui/highgui.hpp>
#include <opencv2/imgproc/imgproc.hpp>
#include <boost/program_options.hpp>
#include <iostream>
#include <vector>
#include <tuple>
#include <chrono>
#include <unistd.h>
#include "PID_controller.h"
#include "camera_controls.h"
#include "async_reader.h"

using namespace cv;
using namespace std;
using namespace std::chrono;
using namespace Capstone;

cv::Rect find_object(cv::Mat tempFrame);
std::tuple<float, float> errors(std::tuple<int, int> center, float width, float height);
size_t limit(float val, size_t max);
std::tuple<int, int> control(const std::tuple<float, float>& errors, PIDController& panPID, PIDController& tiltPID, PTZOptics20x& ptz);
std::tuple<int, int> computeCenter(cv::Rect& boundingBox);

int main(int argc, char** argv )
{
    // Command-line arguments
    namespace po = boost::program_options;
    po::options_description desc("Allowed options");
    desc.add_options()  // returns special proxy object that defines operator()
        ("help,h", "produce help message")
        ("host,H", po::value<string>()->default_value("192.168.1.40"), "Host address for the camera for PTZ control")
        ("port,p", po::value<string>()->default_value("1259"), "Port for TCP comms PTZ control")
        ("dev,d", po::value<int>()->default_value(0), "Camera USB device location for OpenCV")
        ("count,c", po::value<int>()->default_value(600), "Number of frames to run for non-GUI mode")
        ("gui,g", po::value<bool>()->default_value(true),"GUI mode");

    po::variables_map vm;
    po::store(po::parse_command_line(argc, argv, desc), vm);
    po::notify(vm);

    if (vm.count("help")) {
        cout << desc << "\n";
        return 1;
    }

    bool guiMode = vm["gui"].as<bool>();

    string host = vm["host"].as<string>();
    string port = vm["port"].as<string>();

    std::cout << host << " " << port << '\n';
    PTZOptics20x ptz(host, port);
    ptz.connect();

    int width = 1280;
    int height = 720;
    int device = vm["dev"].as<int>();
    Camera source(device, width, height);

    if (guiMode) {
        cv::namedWindow("cam", WINDOW_AUTOSIZE);
    }

    PIDController panPID(1.4, 0.0, 0.0, 1.0/50.0, 2.0*3.14*10.0);
    PIDController tiltPID(1.4, 0.0, 0.0, 1.0/50.0, 2.0*3.14*10.0);

    milliseconds start = duration_cast< milliseconds >(system_clock::now().time_since_epoch());
    milliseconds end;

    while (true) {
        cv::Mat frame = source.read();
        if (frame.empty()) { continue; }

        cv::Rect boundingBox = find_object(frame);

        if (!boundingBox.area()) {
            if (guiMode) {
                cv::imshow("cam", frame);
                char key = waitKey(1);
                if (key == 'q') { break; }
            }
            continue;
        }

        if (guiMode) {
            cv::rectangle(frame, boundingBox, cv::Scalar(0, 255, 255));
            cv::imshow("cam", frame);
            char key = waitKey(1);
            if (key == 'q') { break; }
        }

        std::tuple<int, int> center = computeCenter(boundingBox);

        std::tuple<int, int> speeds = control(errors(center, width, height), panPID, tiltPID, ptz);

        end = duration_cast< milliseconds >(system_clock::now().time_since_epoch());
        float dur = (end - start).count();  // duration in milliseconds
        int FPS = 1/(dur/1000);
        std::cout << "Approx. Frame Rate: " << FPS << '\n';
        start = duration_cast< milliseconds >(system_clock::now().time_since_epoch());
    }

    source.stop();
    source.release();
    if (guiMode) {
        cv::destroyAllWindows();
    }
    ptz.stop();
    usleep(1000000);
    std::cout << "Program completed!" << '\n';
}

cv::Rect find_object(cv::Mat tempFrame) {
    cv::Mat frame = tempFrame.clone();
    int low_H = 100;
    int low_S = 50;
    int low_V = 50;
    int high_H = 160;
    int high_S = 255;
    int high_V = 255;

    cv::blur(frame, frame, cv::Size(11,11));
    cv::cvtColor(frame, frame, COLOR_BGR2HSV);
    cv::inRange(frame, cv::Scalar(low_H, low_S, low_V), cv::Scalar(high_H, high_S, high_V), frame);
    cv::erode(frame, frame, cv::Mat()); cv::erode(frame, frame, cv::Mat());
    cv::dilate(frame, frame, cv::Mat()); cv::dilate(frame, frame, cv::Mat());

    vector<vector<cv::Point>> contours; // Vector for storing contour
    cv::findContours(frame, contours, CV_RETR_EXTERNAL, CV_CHAIN_APPROX_SIMPLE);
    if (contours.size() == 0) { return cv::Rect(); }
    int largest_area = 0;
    size_t largest_contour_index = 0;
    for(size_t index = 0; index < contours.size(); index++) {
        double area = cv::contourArea(contours[index], false); //  Find the area of contour
        if (area > largest_area) {
            largest_area = area;
            // Store the index of largest contour
            largest_contour_index = index;
        }
    }
    // Find the bounding rectangle for biggest contour

    return boundingRect(contours[largest_contour_index]);
}

std::tuple<float, float> errors(std::tuple<int, int> center, float width, float height) {
    float centerX = std::get<0>(center);
    float centerY = std::get<1>(center);
    float pan_error = (width/2 - centerX)/50;
    float tilt_error = (height/2 - centerY)/30;
    return std::make_tuple(pan_error, tilt_error);
}

size_t limit(float val, size_t max) {
    if (abs(val) > max) { return max; }
    else { return (size_t)abs(val); }
}

std::tuple<int, int>
control(const std::tuple<float, float>& errors, PIDController& panPID, PIDController& tiltPID, PTZOptics20x& ptz) {

    unsigned int dur = 10;

    // std::cout << "Error: " << std::get<0>(errors) << ", " << std::get<1>(errors) << '\n';

    float pan_command = panPID.compute(std::get<0>(errors));
    float tilt_command = tiltPID.compute(std::get<1>(errors));

    // std::cout << "pan_command: " << pan_command << '\n';
    // std::cout << "tilt_command: " << tilt_command << '\n';

    size_t pan_speed = limit(pan_command, 24);
    size_t tilt_speed = limit(tilt_command, 20);

    // std::cout << "Pan Speed: " << pan_speed << '\n';
    // std::cout << "Tilt Speed: " << tilt_speed << '\n';

    if ((pan_speed == 0) and (tilt_speed == 0)) {
        ptz.leftUp(1, 1);
        usleep(dur);
    } else if (pan_speed == 0) {
        if (tilt_command == abs(tilt_command)) {
            ptz.up(tilt_speed);
            usleep(dur);
        } else {
            ptz.down(tilt_speed);
            usleep(dur);
        }
    } else if (tilt_speed == 0) {
        if (pan_command == abs(pan_command)) {
            ptz.left(pan_speed);
            usleep(dur);
        } else {
            ptz.right(pan_speed);
            usleep(dur);
        }
    } else if ((abs(pan_command) == pan_command) && (abs(tilt_command) == tilt_command)) {
        ptz.leftUp(pan_speed, tilt_speed);
        usleep(dur);
    } else if ((abs(pan_command) == pan_command) && (abs(tilt_command) != tilt_command)) {
        ptz.leftDown(pan_speed, tilt_speed);
        usleep(dur);
    } else if ((abs(pan_command) != pan_command) && (abs(tilt_command) == tilt_command)) {
        ptz.rightUp(pan_speed, tilt_speed);
        usleep(dur);
    } else if ((abs(pan_command) != pan_command) && (abs(tilt_command) != tilt_command)) {
        ptz.rightDown(pan_speed, tilt_speed);
        usleep(dur);
    }
    return std::make_tuple(pan_speed, tilt_speed);
}

std::tuple<int, int> computeCenter(cv::Rect& boundingBox) {
    int centerX = boundingBox.x + (boundingBox.width/2);
    int centerY = boundingBox.y + (boundingBox.height/2);
    return std::make_tuple(centerX, centerY);
}

File: depr/camera/color_tracker.py
import cv2
import imutils
from time import sleep, time
from camera_controls import PTZOptics20x
from async_reader import Camera
from controller import PIDController, control, limit, errors_pt
import argparse

parser = argparse.ArgumentParser()
parser.add_argument('-H', '--host', default='192.168.1.40', help="Host address for the camera for PTZ control")
parser.add_argument('-p', '--port', default=5678, help="Port for TCP comms PTZ control", type=int)
parser.add_argument('-d', '--dev', default=0, help='Camera USB device location for OpenCV', type=int)
parser.add_argument('-c', '--count', default=600, help='Number of frames to run for non-GUI mode', type=int)
parser.add_argument('-g', '--gui', help='GUI mode', action='store_true')
args = parser.parse_args()

def find_object(frame):

    lower = (100, 50, 50)
    upper = (160, 255, 255)

    blurred = cv2.GaussianBlur(frame, (11, 11), 0)
    hsv = cv2.cvtColor(blurred, cv2.COLOR_BGR2HSV)

    mask = cv2.inRange(hsv, lower, upper)
    mask = cv2.erode(mask, None, iterations=2)
    mask = cv2.dilate(mask, None, iterations=2)

    cnts = cv2.findContours(mask.copy(), cv2.RETR_EXTERNAL, cv2.CHAIN_APPROX_SIMPLE)
    cnts = cnts[0] if imutils.is_cv2() else cnts[1]
    center = None

    if len(cnts) > 0:
        c = max(cnts, key=cv2.contourArea)
        ((x, y), radius) = cv2.minEnclosingCircle(c)
        M = cv2.moments(c)
        center = (int(M["m10"] / M["m00"]), int(M["m01"] / M["m00"]))

        if radius > 10:
            return [center, x, y, radius]

def control_zoom(error, zoom_pid, ptz):

    dur = 0.001

    zoom_command = zoom_pid.compute(error) # positive means zoom in
    zoom_speed = limit(zoom_command, 1)

    if not zoom_speed:
        ptz.zoomstop()
        sleep(dur)

    if zoom_command > 0:
        ptz.zoomin(zoom_speed)
        sleep(dur)
    elif zoom_command < 0:
        ptz.zoomout(zoom_speed)
        sleep(dur)

    return zoom_speed

def error_zoom(size, height):
    target_size = float(height)/3.0  # no specific reason for 3
    return (target_size - size)/30  # no specific reason for 30

gui_mode = args.gui

ptz = PTZOptics20x(args.host, args.port).init()

width = 1280
height = 720
source = Camera(args.dev, width, height)

if gui_mode:
    cv2.namedWindow("cam")
    cv2.moveWindow("cam", 20, 20)

pan_pid = PIDController(1.2, 0.1, 0.1, 1/50, 2*3.14*10)
tilt_pid = PIDController(1.2, 0.1, 0.1, 1/50, 2*3.14*10)
zoom_pid = PIDController(1.2, 0.1, 0.1, 1/50, 2*3.14*10)

start_time = time()

counter = 0
count_limit = args.count

while True:
    frame = None
    frame = source.cvreader.Read()

    if frame is None:
        continue

    out = find_object(frame)

    if out is None:
        if gui_mode:
            cv2.imshow("cam", frame)
            key = cv2.waitKey(1) & 0xFF
            if key == ord('q'):
                break
        continue

    center, x, y, radius = out

    if gui_mode:
        cv2.circle(frame, (int(x), int(y)), int(radius), (0, 255, 255), 2)
        cv2.circle(frame, center, 5, (0, 0, 255), -1)

        cv2.imshow("cam", frame)
        key = cv2.waitKey(1) & 0xFF
        if key == ord('q'):
            break

    pan_tilt_error = errors_pt(center, width, height)
    zoom_error = error_zoom(2*radius, height)
    control(pan_tilt_error, pan_pid, tilt_pid, ptz)
    control_zoom(zoom_error, zoom_pid, ptz)

    print("Frame Rate:", int(1/ (time() - start_time)))

    start_time = time()
    counter += 1

    if not gui_mode and counter > count_limit:
        break

source.cvreader.Stop()
source.cvcamera.release()

if gui_mode:
    cv2.destroyAllWindows()

ptz.stop()
sleep(1)
print("Program complete. Exiting..")
